V4.2 Porting

sgbihu opened the pull request:
Porting V4.2 to improve the performance. Please don't merge.

sgbihu committed:
Fix Jira 768 BLE Appearance Value is 0

Solution:
  Add service and appearance characteristic in the BLE lib.

Changed files
  BLEDevice.cpp - check the local device
  BLEDeviceManager.cpp - Change the calls
  BLEDeviceManager.h - remove appearance member
  BLEProfileManager.cpp - Add GAP service and appearance characteristic
  BLEProfileManager.h - Add set/get API

sgbihu committed:
Jira 714 Find the limits of the number of Services and Characteristics

Make the BLE support more characteristics and services by separate the
  service registration

Changed files
  BLEProfileManager.cpp - Remove register all service at same time
  BLEProfileManager.h - Change the comments
  BLEServiceImp.cpp - Add the register service after update the characteristic

sgbihu committed:
Fix GIT #516 and 517 return success when respond with error

Root cause
1. The return value indecate the read/write request send successfully.
2. The subscribe will ignore the response and this will not be fixed.

Solution
1. Add the error detect based on read/write response.

Changed files
BLECallbacks.cpp - Delete the error check for read response.
BLECharacteristicImp.cpp
  - Add the error check.
  - Optimize the memory allocate and clear the allocated memory.
BLECharacteristicImp.h - Add varibles to flag the response error

sgbihu committed:
Jira 913 BLESubscribed, BLEUnsubscribed Event Handlers are not called

1. Implement the subscribe changed event notify feature.

Changed files:
  BLECallbacks.cpp - Add subscribe changed handler and disconnect process
  BLECallbacks.h - Callback declaration
  BLECharacteristicImp.cpp - Add the event changed handler
  BLECharacteristicImp.h - Handler declaration

sgbihu committed:
Update to Thunderdome V4.2

sgbihu committed:
Refact the code and fix some bugs

1. Subscribe/unsubscribe event cause notification can't send out
2. Delete unused code
3. Update the library.

sgbihu committed:
Refact the code to make it more readable

sgbihu committed:
Fix Jira 834 Unexpected increase in compiled sketch image size

1. Reconstruct the initial process to avoid call BLE API

Changed files
BLEDeviceManager.cpp - Change the init order
  ble_client.c - Add new call
  ble_service.h - Add new API declaration
  ble_service.c - Modify init logic
  ble_service_api.c - Delete unsed file
  nble_driver.c - Modify init logic

sgbihu committed:
Discover Services 1800 and 1801 by pass parameter

1. Fix Jira 769/git #378. Discover GAP service by parameter.
2. Fix Jira 770/git #379 to show the device name in central when discover
    GAP service.

Changed service
  BLEDevice.cpp - Modify the API
  BLEDevice.h - Modify the API
  BLEDeviceManager.cpp - Add get device name logic
  BLEProfileManager.cpp - Implement get devicename and dicovery logic
  BLEProfileManager.h - Add API to get devicename from characteristic
  BLEServiceImp.cpp - Modify the filter logic
  BLEServiceImp.h - Modify the API

sandeepmistry commented on libraries/CurieBLE/src/internal/BLEDeviceManager.h:
```diff
@@ -392,7 +400,7 @@ class BLEDeviceManager
     uint8_t    _peer_adv_data_len[BLE_MAX_ADV_BUFFER_CFG];
     uint8_t    _peer_scan_rsp_data[BLE_MAX_ADV_BUFFER_CFG][BLE_MAX_ADV_SIZE];
-    int8_t     _peer_scan_rsp_data_len[BLE_MAX_ADV_BUFFER_CFG];
+    uint8_t    _peer_scan_rsp_data_len[BLE_MAX_ADV_BUFFER_CFG];
```
@SidLeung @sgbihu I think PR https://github.com/01org/corelibs-arduino101/pull/500 was reverted here. Is this change intended?

sgbihu replied:
This change has been considered. You implementation has an assumption, the scan response should followed by advertisement. But this may has some corner case to make ADV not be reported to sketch.
Now I use timeout mechanism. The central will report ADV data after 800ms delay.

sgbihu committed:
Fix Jira 903 BLE peripheral.characteristic() sometimes returns NULL unexpected

Root cause:
1. The stack doesn't release the resource and makes peripheral crashed.
    Then central blocked.

Solution:
1. Fix memory leak issue and release the resource.

Changed File:
  gatt.c - Modify the code to make sure the resource can be released.

sgbihu committed:
Fix Jira 925 KW errors found in PR534

1. Dismiss the warnings.
2. Rebuild the library

Changed files:
  BLEProfileManager.cpp - Init member variable
  rpc_deserialize.c - Delete unreachable code
  dtm.c - Init the variables
  gap.c - Init the variables
  gatt.c - Init the variables

sgbihu committed:
Delete unsed logs and make avaliable more fast

sgbihu committed:
Fix BLE crash issue when do long time test

1. Reinitial the BLE MAC address to avoid BLE start timer

sgbihu committed:
Fix Jira 923 BLE 4.2 BLECharacteristic::setValue() hangs

Root cause
  1. The central powered off peripherial device and make it not give response.

Solution
  1. Exit the wait loop when connection lost.

Changed file
  BLECharacteristicImp.cpp - Change the wait logic

sgbihu committed:
Add comment for KW warning issue

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/arduino/ArduinoCore-arc32?pullRequest=534) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/arduino/ArduinoCore-arc32?pullRequest=534) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/arduino/ArduinoCore-arc32?pullRequest=534) it.</sub>
